File: leetcode/medium/378_kthSmallest.ts
function kthSmallest_0(matrix: number[][], k: number): number {
    const arr: number[] = matrix.reduce((acc, row) => acc.concat(row), []);
    arr.sort((a, b) => a - b);
    return arr[k-1];
};

function kthSmallest(matrix: number[][], k: number): number {
    let count: number = 0;
    let row: number = 0;
    let col: number = 0;

    while (count < k) {
        if (count + 1 === k) {
            return matrix[row][col];
        }
    }

    return -1;
};

/*

1, 2, 3, 10,
2, 3, 4, 11
5, 6, 7, 12
10,11,12,13

*/

const test = () => {
    const params = [
        {
            input: {
                matrix: [
                    [1, 5,  9],
                    [10,11,13],
                    [12,13,15]
                ],
                k : 8,
            },
            output: 13,
        },
        {
            input: {
                matrix: [[-5]], k: 1
            },
            output: -5,
        },
    ];

    params.forEach(({input, output}) => {
        const { matrix, k } = input;
        const result = kthSmallest(matrix, k);

        console.log(
            JSON.stringify(result) === JSON.stringify(output) ? 'SUCCESS ' : 'ERROR ',
            'input', JSON.stringify(input),
            'output', output,
            'result', result,
        );
    });
};

test();

File: leetcode/medium/378_kthSmallest.py
from typing import List
import json
from collections import deque, defaultdict
import heapq


class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        N = len(matrix)
        merged = []

        for row in matrix:
            i = 0
            j = 0
            current = []
            while i < len(merged) and j < len(row):
                if merged[i] <= row[j]:
                    current.append(merged[i])
                    i += 1
                else:
                    current.append(row[j])
                    j += 1

            while i < len(merged):
                current.append(merged[i])
                i += 1

            while j < len(row):
                current.append(row[j])
                j += 1

            merged = current


        return merged[k-1]





def test ():
    params = [
        {
            'input': [[[1,5,9],[10,11,13],[12,13,15]], 8],
            'output': 13,
        },
        {
            'input': [[[-5]], 1],
            'output': -5,
        },
    ]
    solution = Solution()

    for param in params:
        matrix, k = param['input']
        result = solution.kthSmallest(matrix, k)
        correct = json.dumps(result) == json.dumps(param['output'])

        msg = 'SUCCESS' if correct else 'ERROR'
        msg += '\n'
        if not correct:
            msg += 'input ' + json.dumps(param['input']) + '\n'
            msg += 'output ' + json.dumps(param['output']) + '\n'
            msg += 'result ' + json.dumps(result) + '\n'

        print(msg)


if __name__ == '__main__':
    test()
